### DISCUSSION TITLE
Zero-Copy transfer support for non-plain [13306]

### DISCUSSION THREAD

**Discussion by handler1224mother** [eProsima/Fast-DDS#2393] [Q&A] [answered]
is the latest release of fast-dds support Zero-Copy transfer support for non-plain?

**Comment by handler1224mother** [comment]
when i use take _instance for non-plain type data, after a few moment the subscriber stop the callback of on_data_available.
can you help me to solve it ? thank you very much

**Comment by EduPonz** [comment, marked as answer]
Hi @handler1224mother ,

> is the latest release of fast-dds support Zero-Copy transfer support for non-plain?

Unfortunately, the loaning APIs (Zero-copy is a combination of Data sharing and loans in both DataWriter and DataReader) only support plain types (see [this](https://fast-dds.docs.eprosima.com/en/latest/fastdds/use_cases/zero_copy/zero_copy.html#constraints)).

> when i use take _instance for non-plain type data, after a few moment the subscriber stop the callback of on_data_available.
> can you help me to solve it ? thank you very much

Yes, you should be able to use `DataReader::take_instance` for non-plain data as well. What do you mean by `the subscriber stop the callback of on_data_available`? I'm afraid we would need more information to help you with your problem.

In any case, if these are two separe questions, I'd appreciate if you could open two different tickets to help us track them better.

**Comment by xixioba** [comment]
Does Zero-Copy support **sequence**?

**Reply by JLBuenoLopez** [reply]
#3810

**Comment by zoeeer** [comment]
What are the reasons for the plain and bounded type constraints? I suspect it might be to facilitate the use of a Pre-allocated PayloadPool? It seems to me that implement loaning is also possible for dynamically allocated PayloadPools (i'm not so sure). Or maybe there are other reasons for the constraints?

---- edit ----
I see here [the doc](https://fast-dds.docs.eprosima.com/en/latest/fastdds/use_cases/zero_copy/zero_copy.html#constraints) explains why we need plain types (to avoid copy on serialization). So my question would be more about the "bounded" constraint.

**Reply by JesusPoderoso** [reply]
@zoeeer A bounded type is that which has a limit. For instance, IDL syntax allows to define a sequence, string and map without specifying the maximum number of elements of the container. Thus, the type is considered unbounded if the optional upper limit parameter is not defined.
As  @EduPonz said, Zero-copy is a combination of Data sharing and loans in both DataWriter and DataReader.
The bounded constraint is imposed by data-sharing mechanism to avoid the encapsulation, so the serialized data is directly written in the segment; meanwhile the plain constraint is imposed by the loans (if used in both DataWriter and DataReader).
If the type is **NOT** bounded,  it cannot be plain. So meeting the plain constraint also meets the bounded constraint.
